examples/features/torch_batch_process/batch_inference/compare_with_core_api/core_api_inference.py

```diff
@@ -0,0 +1,151 @@
+            X = next(dataloader_iterator, None)
+            logging.info(f"Working on batch is {batch_idx}")
+            if X is not None:
+                data, label = X
+                output = model(data)
+                preds = output.argmax(dim=1, keepdim=True)
+
+                file_name = f"inference_out_{rank}_{batch_idx}.json"
+
+                output = []
+
+                for pred in preds:
+                    output.append(pred[0].item())
+
+                with open(
+                    os.path.join(inference_output_dir, f"{file_name}"),
+                    "w",
+                ) as f:
+                    json.dump({"predictions": output}, f)
+
+                # After each batch, synchronize and update number of catches completed
+                if context.distributed.rank == 0:
+                    work_completed_this_round = sum(context.distributed.gather(len(data)))
+                    records_processed += work_completed_this_round
+                    checkpoint_metadata = {
+                        "steps_completed": batch_idx,
+                    }
+                    with context.checkpoint.store_path(checkpoint_metadata) as (path, uuid):
+                        with open(os.path.join(path, "batch_completed.json"), "w") as file_obj:
+                            json.dump({"batch_completed": batch_idx}, file_obj)
+                else:
+                    context.distributed.gather(len(data))
+
+                if context.preempt.should_preempt():
+                    return
+
+
+def _get_data_loader(batch_size: int, total_worker: int, rank: int, skip: int) -> [Any, int]:
+    transform = transforms.Compose(
+        [transforms.ToTensor(), transforms.Normalize((0.5, 0.5, 0.5), (0.5, 0.5, 0.5))]
+    )
+    with filelock.FileLock(os.path.join("/tmp", "inference.lock")):
+        inference_data = tv.datasets.CIFAR10(
+            root="/data", train=False, download=True, transform=transform
+        )
+    dataloader = pytorch.DataLoader(
+        dataset=inference_data,
+        batch_size=batch_size,
+        shuffle=False,
+    ).get_data_loader(repeat=False, skip=skip, num_replicas=total_worker, rank=rank)
+
+    # Enumerate over dataloader directly may cause some workers to iterate for 1 more time
+    # than others when drop_last = False. If those workers synchronize on the last batch_idx,
+    # they would hang forever as other workers never hit that last batch_idx.
+    # To avoid the issue, we calculate and take the ceiling of the iteration count to ensure
+    # all workers iterate for the same number of times.
+    per_worker_iterate_length = math.ceil(len(inference_data) / batch_size / total_worker)
+
+    return dataloader, per_worker_iterate_length
+
+
+def _load_state(checkpoint_directory: str):
+    checkpoint_directory = pathlib.Path(checkpoint_directory)
+    with checkpoint_directory.joinpath("metadata.json").open("r") as f:
+        metadata = json.load(f)
+        return metadata
```

Comment:
Nit: read from `batch_completed.json` instead.

It happens to be true that in this instance, the one required piece of metadata matches exactly the one piece of state you need to save.

But that's not generally the case.  Generally, the files of the checkpoint should be the state you are preserving between runs.  Generally, metadata is additional, mutable data _about_ those checkpoint files.  It happens that we save a copy of the metadata as metadata.json... but it would be cleaner to treat that as informational metadata, and read from `batch_completed.json`.

Does that make sense?

Reply:
Yes, per later comment, we are changing to storing just the steps_completed directly, will read from that record directly.